### ISSUE TITLE
TurnFunction - Difference between 4.9.1 and Master

### ISSUE THREAD

**Issue by OnlyMate** [Project-OSRM/osrm-backend#4037]
Hey! 

I'm trying to update our server to the master version and struggling at the turn function.

I'm wondering why the angle of the _turn_function_ Parameter is different.

Here is the 4.9.1 turn function:

```lua
function turn_function (angle)
	-- compute turn penalty as angle^2, with a left/right bias
	k = turn_penalty / (90.0 * 90.0)
	if angle >= 0 then
		return angle * angle * k / turn_bias
	else
		return angle * angle * k * turn_bias
	end
end
```

Here is the Master turn function:

```lua
function turn_function(turn)
  local k = turn_penalty / (90.0 * 90.0)
  if turn.angle >= 0 then
    turn.duration = turn.angle * turn.angle * k / turn_bias
  else
    turn.duration = turn.angle * turn.angle * k * turn_bias
  end
  turn.weight = turn.duration 
end
```

_angle_ and _turn.angle_ are different. I got the same dataset. Can you tell me why the parameter isn't the same?

**Comment by MoKob** [comment]
@OnlyMate OSRM is trying it's best to approximate a good turning angle that represents the road we are turning onto.

<img width="452" alt="screen shot 2017-05-12 at 14 46 50" src="https://cloud.githubusercontent.com/assets/12932279/25998618/26a43c70-3722-11e7-8185-bb45a4957d51.png">

In situations like the above picture, the road looks like its taking a slight turn to the right. Accounting for lanes and/or the split into dual carriage ways, the path from left to right doesn't take a turn.

There are many situations were OSM models road in specific structures that accurately represent the way but not represent how the angle is perceived.
If you see different angles, it is due to differences in the turn angle calculation which only looked at the very first coordinate along the way which is very prone to errors due to tiny offsets / skipping over lanes on exits and similar settings.

**Comment by OnlyMate** [comment]
@MoKob thank you for your answer. 

I found the old angle calculation here:
https://github.com/Project-OSRM/osrm-backend/blob/v4.9.1/util/compute_angle.cpp

Can you tell me where I find the new angle calculation? 
I was just surprised why I get a very different angles for the same roads.

**Comment by MoKob** [comment]
What you linked to is the angle between coordinates. The coordinates passed to that position may differ though.

We compute the intersections as shapes, which represents a top down view of how we perceive turn angles. The coordinate_extractor is called from [here](https://github.com/Project-OSRM/osrm-backend/blob/e064a9334bbd022365ad1c5c50f25b8093fb72e7/src/extractor/guidance/intersection_generator.cpp#L93-L101), it's source can be found [here](https://github.com/Project-OSRM/osrm-backend/blob/e064a9334bbd022365ad1c5c50f25b8093fb72e7/src/extractor/guidance/coordinate_extractor.cpp).

Intersections are further normalised for dual carriage-ways effects in [here](https://github.com/Project-OSRM/osrm-backend/blob/e064a9334bbd022365ad1c5c50f25b8093fb72e7/src/extractor/guidance/intersection_normalizer.cpp#L35-L40).

**Comment by OnlyMate** [comment]
I posted this link because it is used as the angle for the Lua _turn_function_. 
The _turn_penalty_ is calculated with this angle [here](https://github.com/Project-OSRM/osrm-backend/blob/v4.9.1/extractor/edge_based_graph_factory.cpp#L508-L511). And the calculated _angle_ is given to the Lua-Function [here](https://github.com/Project-OSRM/osrm-backend/blob/v4.9.1/extractor/edge_based_graph_factory.cpp#L635-L636).

In the master version I found [this](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/edge_based_graph_factory.cpp#L509-L516) and where the [_turn_object_](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/scripting_environment_lua.cpp#L597-L605) is given to the Lua-Function. But not where the _angle_ of the _turn_object_ is calculated. 

I'll take a look at your posted links, thanks!

**Comment by daniel-j-h** [comment]
I would go up from [here](https://github.com/Project-OSRM/osrm-backend/blob/e064a9334bbd022365ad1c5c50f25b8093fb72e7/src/extractor/edge_based_graph_factory.cpp#L509) on.

**Comment by OnlyMate** [comment]
After a few debug sessions I got the position where the turn angle is calculated:
The first position is [here](https://github.com/Project-OSRM/osrm-backend/blob/master/src/util/coordinate_calculation.cpp#L128-L154). But I don't know which coordinates are used here. Could you explain that to me? 
This calculated value is used [here](https://github.com/Project-OSRM/osrm-backend/blob/e6ff17ab2ac37518b9da108191d3aad2dce5d1b5/include/util/bearing.hpp#L121-L132). 
The _(180-angle)_ is given to the Lua-Function. Am I right?

**Comment by MoKob** [comment]
@OnlyMate as I've already pointed out, the turn angle calculation does not change looking back quite a few versions. Only the selection of coordinates does. I've linked you to our [coordinate extractor](https://github.com/Project-OSRM/osrm-backend/blob/e064a9334bbd022365ad1c5c50f25b8093fb72e7/src/extractor/guidance/coordinate_extractor.cpp) already. The coordinate extractor tries to find representative coordinates for noisy input.

**Comment by danpat** [comment]
@OnlyMate To be clear - OSRM doesn't just use the first coordinate after an intersection to calculate the angle.  The code that @MoKob linked to is where we try to calculate the best angle that a human will perceive.

Take a look at some of the examples in the [original pull request](https://github.com/Project-OSRM/osrm-backend/pull/2779) that implemented this logic - you'll see that OSRM 4.9.1's logic of simply using the first coordinate after an intersection breaks down in many cases - OSRM 5.x tries to do a better job of giving a "reasonable" angle for scenarios where we previously gave crazy results.

**Comment by OnlyMate** [comment]
Okay, got it now. Thank you all!
